### ISSUE TITLE
Wrong hit result

### ISSUE THREAD

**Issue by aureole82** [paperjs/paper.js#819]
I got the hit result `{ type: 'fill', item: Path@1 } ` where definitely no hit result should be.

![wrong hittest](https://cloud.githubusercontent.com/assets/4819891/10937413/ef3c8d08-82f2-11e5-9495-77b47f9fd46b.png)

http://jsfiddle.net/x5t1v10e/

**Comment by iconexperience** [comment]
Oh dear, that seems to be a side effect of the new boolean code. We probably get two intersections at the little horizontal line left to the point, which then results in a winding number of 1, which means we are inside the path.

**Comment by lehni** [comment]
Hmm but we didn't change the winding code, did we? The `contains()` code doesn't check for intersections. But that's easy verified by comparing with an older version of the library...

**Comment by iconexperience** [comment]
You are correct @lehni, this also happens in older versions, but it is a bit difficult to compare, because in previous versions the curve values were totally wrong (including a few `NaN`).

As I see it, there are a two strange things happening here:

- If I set the stroke width to zero, we still get a hit, because for both vertical lines at the right (the one from [840, 0] to [840, 20] and the one from [860, 20] to [860, 110] the winding numbers are counted in `getWinding()`, so the algorithm thinks we are crossing the path twice there. This is probably what causes the incorrect hit result.

- Curve.solveCubic has not a very good precision for a vertical linear curve in the example below. This is probably not a problem, because we are withing our curve time tolerance. But I find it odd.
```javascript
var values = [0, 0, 0, 0, 0, 20, 0, 20];
var roots = [];
paper.Curve.solveCubic(values, 1, 20, roots, 0, 1);
console.log(roots[0]; // should be "1", but actually is "0.999999988079071"
```

**Comment by iconexperience** [comment]
What causes the double counting of the winding is the condition in <a href="https://github.com/paperjs/paper.js/blob/develop/src/path/PathItem.Boolean.js#L361"> `getWinding()` at line 361</a>, that only omits double counting if there are no horizontal curves with `winding == 0` inbetween.

@lehni Do you know any cases where this condition is required? I commented it out and in my test suite it caused no new glitches.

**Comment by iconexperience** [comment]
I think an easy solution would be the following, but it is a bit of a hack, because `prevT` does not refer to `prevCurve` after this special case. But they are not used in combination, so it may be alright.

In <a href="https://github.com/paperjs/paper.js/blob/develop/src/path/PathItem.Boolean.js#L396">`getWinding()` line 396</a> add this:

```javascript
} else if (winding === 0 && py === values[1] && py === values[7]) {
    prevCurve = curve;
}
```

**Comment by lehni** [comment]
@iconexperience the code in question was implemented to solve #736, so any changes to it should probably be tested against that case as well. Could you have a look?

**Comment by iconexperience** [comment]
Yes, I will do. Also, I think the issue is a little bit more complicated than I thought and there could be other cases that still fail. It should be solveable, but it's a bit tricky.

**Comment by iconexperience** [comment]
I found some inconsistencies with precision in `getWinding()`. In <a href="http://sketch.paperjs.org/#S/rVRRT8IwEP4rR5+2gGMSeQF9MMQHE4NEfGM8jO3YKqUlXaeJhP/udZuAyNAQlqxLr3fffffddWsmwyWyHhsv0EQpa7FIxXb/HmpIuRkpLk0GdzAJJNAj8QNW4Qq1V5w4112/BbS4LWi3gcuMxwgaI3PK3a/clQSjVjWe/m9PEDg3ECktUZ+K6u6ibEStK1zBMF+i5lEovIfR+PHpebgX/pZnBlRuiprqgToE1Pw/kOZJ+oc6x3gdQarXzuLU0DqCM1PGqGUgp/1ABpJOX6h/oUwEenAvBKzKGchSlYuYEmWZnQwwmFEddk5sv2lELI1RaFJvG+8UpsN+WuOYf+LO5lJmC+LNuRADJZQmuIBpjANGR3PaOzYRJ7Pfp8/tbjY9gTIxKVmbTRfWpSKRkpki/kIljsUhfgEjOTi9AeuVm0ajyElIr1SKs0Wc8GnBaGPlKAQZcB19qzHLTakIdMA5GEr3hEhRgbEvU4m6r9HuMtFCDMqYA1VmIseLyFIxOiZMlfi0NHZSlAkNxlVx5wtUxesCzrnpupfo+g9u55dJ/8SZxnBRXgPWm0w3Xw==">this Sketch</a> you can see that points are counted as being on the path if they are horizontally just outside the path (epsilon = 2e-7 is used), but we do not allow this tolerance vertically.
We may want to take a look at the vertical tolerance after this issue has been solved.

**Comment by iconexperience** [comment]
I have created a small test case based on @aureole82's example that produces six wrong results. 

![image](https://cloud.githubusercontent.com/assets/2122773/10996087/3badb6ea-8482-11e5-90ec-b706c168fa7b.png)

For the points outside the path `contains()` returns `true`, for the points inside the path it returns `false`

<a href="http://sketch.paperjs.org/#S/nZPPcoIwEMZfJZNeYGQUtdqq7ckXcNqDB+AAZJWMaeKEqAfHd+8GQfl3aZkJIb/dbD6+nVypjH+ALun3AUyaUY+mitn1OdbkGJtso7g0OfkkQahDSfAJ3n2PjH0/8lqkBjDcRyYNMuklr51dXTJvkFmHTHpJU/Pc76BZSSyIVvZtR+EEGiDhQjboiPO0xS2SjsMdF2KthNKYFtIXAAhpGcqNVgeoBZMkaQW3nJkMg+OSpkLlwBAYfYIqE8yWS8bl/uskwAkpnEEqxkLqPoRWYjNuero2bbu96Ng/a5Np28dFx9i3Gnl6NhoRpuOLlUKOhRZLd+iBYxVyVOavcPp4ih0KkHuTIR0MXHItT7DZKdepgFoHhuuCOMXabnYeVQIeBaiGNMA4cj0yuxuFNe/1Wk3TwMq23Kqf+IdckiqZKywu1N7hZICFlzjwA9uqpIm5zP8i23WbkuwR1Yx3NdEQH+7+0mUQ3X4B">Here is the Sketch</a>

I have set the winding rule to "evenodd", which causes the points inside the path to be counted as outside.

I think I have found a solution, but I need to test it agains some edge cases.

**Comment by iconexperience** [comment]
@lehni I have a question. 
In <a href="https://github.com/paperjs/paper.js/blob/develop/src/path/PathItem.Boolean.js#L357">`getWinding()`line 357</a> there is a check for an "end" of loops, but this is not done by checking if this is the last curve in the array of curves but by `curve.next !== curves[i + 1]`. I do not understand why it is done this way. Do you have any example where there can be several "loops" and where simply checking for the last curve will fail? If there can be several loops I find it strange is that `startCounted` is never reset to `false`, so if the start of one loop is counted, all starts of the following loops are marked as counted.

What I need to do is to find out the first and the last curve that has `winding !== 0 `. For these two curves I then need to make sure that for the start of first and end of last the winding is only added once.

**Comment by lehni** [comment]
The reason why it's done this way is simply that the curves array of CompoundPaths contain more than one loop (all curves of the first child-path = loop 1, followed by all curves of the second child-path = loop 2, etc).

**Comment by lehni** [comment]
@iconexperience your observation about  `startCounted` is probably correct. I need to look into this when I find the time, very busy at the moment unfortunately...

**Comment by iconexperience** [comment]
@lehni I will try to do this. I think I understand the whole algorithm and I know where the problems are. Also, your information about the compound paths was very helpful. But I need some time, too.

**Comment by lehni** [comment]
@iconexperience great, thanks!

**Comment by lehni** [comment]
@iconexperience I've rewritten [your sketch](http://sketch.paperjs.org/#S/lZPLbsIwEEV/xXIXBBFBgIaW0KoLfgC1iy6SLPIYEgvXRo6BBeLfOw6PPDdEchyfmbGvr50zFdEfUI/+7EAnObVpIlMzPkaK7COdbyQTuiCfxA9UIAg+/rtjk6njhHaL1ACG+8isQWa95LVT1SWLBnE7ZNZLmpoXTge5N2JAuDJv00on0AABJ7JBR6zKlmGZtB9vGedryaXCtIC+AEBAb6FCK7mDWjCO41bwl6U6x+D0RhMuC0gRaHWAeyboXyZSJrLvAwcroHAEIdM0oMOH0LvYnOmeU5u33V527HfbZN72cdkx9q1GKs8mE5Kq6GSkkH2pxdAtemAZhQyVOSvsPiqxYw4i0znS0WhIzib/cp/u2UIjzaQnUuiIidJM9PU6KqzyJE2t9ZjEZ6GPuyANMA2HV3fNfFhdSA5jLjOLkREepYcNPx6LVLnl2kwlHGr3ZrwuyTOr28StCSjLG3et2t8XGWQKQAyIRwYK0sGqYaCpv/f4j8cKot31XKjnh5d/) a bit to give visual feedback.

**Comment by iconexperience** [comment]
@lehni I though I had a good solution, but then I found new cases that caused problems. Therefore I will present some intermediate results for discussion.

The problem with the current implementation of `getWinding()` is that it completely ignores curves with `winding==0` (I call them horizontal curves). As a result the merging of intersections if one intersection is at the end point of one curve and another intersection is at the start point of the next curve can fail. Here is an example:

![image](https://cloud.githubusercontent.com/assets/2122773/11038339/bf2f5e92-8702-11e5-9733-11d018e7270e.png)

The two red intersections should be merged, but this fails in the current implementation, because due to the horizontal curve the two vertical curves are not regarded as being neighbors.

I have found a way that fixes merging of these intersections, but my apporach fails if the point is directly on a horizontal curve. Actually, I do not even know what the correct result should be. Below are three examples, red intersections indicate a winding contribution of -1, green ones indicate a winding contribution of 1.

Here is a quick summary of our algorithm which should help to understand the examples: We calculate the sum of the winding contributions for all intersections right and left of the point (`windLeft` and `windRight`). From these sums we calculate the absolute value and use the maximum of both values as the overall winding (`winding = max(abs(windLeft), abs(windRight))`.

A point is inside a path if the winding value is non-zero (when using the non-zero rule) or odd (when using he even-odd rule). 

With the current implementation in  examples 1 and 2 the point will be counted as being inside the path, in example 3 the point is counted as being outside of the path.

**Example 1**
![image](https://cloud.githubusercontent.com/assets/2122773/11037828/2780a530-8700-11e5-85ea-d7a6b09c7ff6.png)
`windLeft = -1, windRight = 0, winding = 1`


**Example 2**
![image](https://cloud.githubusercontent.com/assets/2122773/11037816/183cf196-8700-11e5-9a62-fef1ac2ef50e.png)
`windLeft = 0, windRight = 1, winding = 1`


**Example 3**
![image](https://cloud.githubusercontent.com/assets/2122773/11037797/09090aa2-8700-11e5-881f-29689e65a3bc.png)
`windLeft = 0, windRight = 0, winding = 0`

**Comment by iconexperience** [comment]
For testing my first proposed solution, I have made the test case a little bit nastier.

<a href="http://sketch.paperjs.org/#S/lZTLboMwEEV/xXIXECWikFelpFUX/YGoXWQBLAhMgxXXjozbLKr+e8fOA8hkEyQIHN8ZD3eG/HJVfAFf8I8d2LLmI17qyj3/FIbtC1uvtFC2YS8szUymGB5pMo1HLInjfHSNuiS+ScY9Mr5JpiSKknmXjAnxeSjpFz2JCZqdiAP50l3d6b1ACxQc2Ao9CVtjBl60jz6FlG9aaoOyjD8AQMZPS401egedxc1mc7W4FpWtcTE50VLqBioE1nzDWQl2LVQl1Pb9W0KYcfgBpasq44NLoedia2Fv9G1COpJckwlpSEII1TzdjJpdJ54TyZwk7pMZ6TTVkB7O6XROKHqi49BDbfcfH1llioMzle29q45+YjdD57VAj+Ml/jy3tkcS1NbWSIfDAft1+r9zunsDXW1OXmplC6H8WOCEHJ+a0M+kiw0vSVKRp/gWrAeSfHCcE5cPoxstIZJ6Gwo2xKFc4Ik3l01ard9bmFJC5wuI3jy5Z3cciE4BPrz31bTv98qCrQFQAVuwwEAVLC8GstOB/1MbA8Xu2A++SPO/fw==">Here is the sketch</a>

And this is how it looks:

![image](https://cloud.githubusercontent.com/assets/2122773/11565522/acdf1d24-99de-11e5-89e4-af7c1a12a879.png)

As you can see, the results look quite random.

**Comment by iconexperience** [comment]
@lehni Do you have a proposal what the correct return value of `getWinding()` for the numbered points below is?

![image](https://cloud.githubusercontent.com/assets/2122773/11566733/02076d6a-99e4-11e5-9634-ecb9968eb22a.png)

For 1, 2 and 6 it is probably `1`, but what about 3, 4 and 5? Would it be sufficient if we returned `1`? This would mean *inside the path* for *even-odd* and *non-zero* winding rules.

**Comment by iconexperience** [comment]
<a href="http://sketch.paperjs.org/#S/lVTNbrMwEHwVy98hoEQpJE0qJa166AtU7aH9BBwIbIIV146M0xyqvnttCD9m6aFITmB2xvaOvftFRfoBdENfj6Czgs5oJnP7/ZkqUl6Ywd7/kweyP4tMMym8k2RClz75ilUsiHn2UhHPspmhBVvzd09q0pyDOOjCQNNpT2Afy9cfJ6OoqRFLoiDZ9in9gMMLf+OFlmdmbcPf9qX6aYZd95Tq4rkSGXrUcKPwNpiRMAiS2RDqI8EosnCQxShyi1QYWfeRBUKqeTDibnoZIGh1RSxQm9d4UTA9YsUSJRkOkSXKMUQI5tyNqlbDideIskYTu8gKmYc5yJY1PvAlhu6www7UGdqUi9fdMH/rBFq7fecUbB0IuJBno0Pi03zPOH+S3NTZA4npPwCI6TVUaiWP0AvudrtB8I3lujDB8IpmXJaQ20pRZ2iYoN+YyJk4vJw5eDGFTxAyz2PabXQ4bm5IrtKLvULXGrToWDdosx5pCG11/lXYdJFMCp0yUWVkkqu/Sq+y02o7z+tWMiMOECa+3/YLoy4lhzmXB4+RqfFzY4Z5aRfpuNXaTGUceoc3f6qQv6xurn9vA5XcOfAuv0cyOSgAMSEbMlGQT7atgU0vNO17pyA91udBN1Hy/QM=">Here is another Sketch</a> which shows the results if we switch x and y corrdinates in the example above. This is how it looks:

![image](https://cloud.githubusercontent.com/assets/2122773/11571067/cb2f4116-99fa-11e5-997f-78f7ccc8c952.png)

 I will try to get the same results for the original example.

**Comment by iconexperience** [comment]
I have finished an implementation that seems to work well. 

After some testing and cleanup I hopefully can create a pull request next week.

**Comment by lehni** [comment]
@iconexperience sounds great! Apologies for my silence on all fronts, I've been constantly absorbed, and will be for another two weeks until things are finally cooling down. Will be back more around here after that!

**Comment by iconexperience** [comment]
No problem, I know that you are busy. I will just keep on working and see how far I can get.

**Comment by lehni** [comment]
So I'm finally coming out of my work tunnel. What's the state of affairs here now? Is #852 fully solving this? And could you outline the approach you've settled for in the end? Just trying to understand it.

**Comment by lehni** [comment]
The reason why I am asking: I wonder if a part of the work that you're doing in `getWinding()` now could be moved to `_getMonoCurves()` curves and cached instead?

**Comment by iconexperience** [comment]
As far as I can see this issue is fixed with my proposed changes at #852. I have not found any negative side effects and performance should be about the same. But since the change is rather complex, I would like to do more testing.

And I will add more details on my new implementation here.

**Comment by lehni** [comment]
This is now merged in, but I am keeping this issue open until I've implemented unit tests for it too.
